### PR TITLE
Add Enmap as custom database

### PR DESCRIPTION
**Pull request by GreepTheSheep**
I had tried to do the same system as quick.db when I had Enmap, but I noticed that it was not the same system (the base gets corrupted when adding). So I modified and found this solution for those who use [Enmap](https://enmap.evie.dev/).

### CODE REVIEW

File: examples/custom-databases/enmap.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+const Discord = require('discord.js'),
+    client = new Discord.Client(),
+    settings = {
+        prefix: 'g!',
+        token: 'Your Discord Token'
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:5]:
```suggestion
        token: 'Your Discord Bot Token'
```

File: examples/custom-databases/enmap.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+// Load Enmap
+const Enmap = require('enmap');
+
+// Create giveaways table
+const giveawayDB = new Enmap({name: "giveaways"})
+
+const { GiveawaysManager } = require('discord-giveaways');
+const GiveawayManagerWithOwnDatabase = class extends GiveawaysManager {
+
+    // This function is called when the manager needs to get all the giveaway stored in the database.
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:17]:
```suggestion
    // This function is called when the manager needs to get all the giveaways stored in the database.
```

File: examples/custom-databases/enmap.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        return giveawayDB.fetchEverything().array();
+    }
+
+    // This function is called when a giveaway needs to be saved in the database (when a giveaway is created or when a giveaway is edited).
+    async saveGiveaway(messageID, giveawayData){
+        giveawayDB.set(messageID, giveawayData);
+        return true;
+    }
+
+    async editGiveaway(messageID, giveawayData){
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:28]:
```suggestion
    // This function is called when a giveaway needs to be edited in the database.
    async editGiveaway(messageID, giveawayData) {
```

File: examples/custom-databases/enmap.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        exemptPermissions: [ 'MANAGE_MESSAGES', 'ADMINISTRATOR' ],
+        embedColor: '#FF0000',
+        reaction: '🎉'
+    }
+});
+// We now have a giveawaysManager property to access the manager everywhere!
+client.giveawaysManager = manager;
+
+client.on('ready', () => {
+    console.log('I\'m ready !');
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:55]:
```suggestion
    console.log('I\'m ready!');
```

File: examples/custom-databases/enmap.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+};
+
+// Create a new instance of your new class
+const manager = new GiveawayManagerWithOwnDatabase(client, {
+    storage: false, // Important - use false instead of a storage path
+    updateCountdownEvery: 10000,
+    default: {
+        botsCanWin: false,
+        exemptPermissions: [ 'MANAGE_MESSAGES', 'ADMINISTRATOR' ],
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:46]:
```suggestion
        exemptPermissions: ['MANAGE_MESSAGES', 'ADMINISTRATOR'],
```

File: examples/custom-databases/enmap.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+// Load Enmap
+const Enmap = require('enmap');
+
+// Create giveaways table
+const giveawayDB = new Enmap({name: "giveaways"})
+
+const { GiveawaysManager } = require('discord-giveaways');
+const GiveawayManagerWithOwnDatabase = class extends GiveawaysManager {
+
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:16]:
```suggestion
```

File: examples/custom-databases/enmap.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    // This function is called when the manager needs to get all the giveaway stored in the database.
+    async getAllGiveaways(){
+        return giveawayDB.fetchEverything().array();
+    }
+
+    // This function is called when a giveaway needs to be saved in the database (when a giveaway is created or when a giveaway is edited).
+    async saveGiveaway(messageID, giveawayData){
+        giveawayDB.set(messageID, giveawayData);
+        return true;
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:25]:
```suggestion
        // Don't forget to return something!
        return true;
```

File: examples/custom-databases/enmap.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    async editGiveaway(messageID, giveawayData){
+        giveawayDB.set(messageID, giveawayData);
+        return true;
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:30]:
```suggestion
        // Don't forget to return something!
        return true;
```

File: examples/custom-databases/enmap.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    // This function is called when a giveaway needs to be deleted from the database.
+    async deleteGiveaway(messageID){
+        giveawayDB.delete(messageID);
+        return true;
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:36]:
```suggestion
        // Don't forget to return something!
        return true;
```

File: examples/custom-databases/enmap.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    settings = {
+        prefix: 'g!',
+        token: 'Your Discord Token'
+    };
+
+// Load Enmap
+const Enmap = require('enmap');
+
+// Create giveaways table
+const giveawayDB = new Enmap({name: "giveaways"})
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:12]:
```suggestion
const giveawayDB = new Enmap({ name: 'giveaways' })
```

File: examples/custom-databases/enmap.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+const { GiveawaysManager } = require('discord-giveaways');
+const GiveawayManagerWithOwnDatabase = class extends GiveawaysManager {
+
+    // This function is called when the manager needs to get all the giveaway stored in the database.
+    async getAllGiveaways(){
+        return giveawayDB.fetchEverything().array();
+    }
+
+    // This function is called when a giveaway needs to be saved in the database (when a giveaway is created or when a giveaway is edited).
+    async saveGiveaway(messageID, giveawayData){
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:23]:
```suggestion
    async saveGiveaway(messageID, giveawayData) {
```

File: examples/custom-databases/enmap.js [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        return true;
+    }
+
+    async editGiveaway(messageID, giveawayData){
+        giveawayDB.set(messageID, giveawayData);
+        return true;
+    }
+
+    // This function is called when a giveaway needs to be deleted from the database.
+    async deleteGiveaway(messageID){
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:34]:
```suggestion
    async deleteGiveaway(messageID) {
```

File: examples/custom-databases/enmap.js [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+const Enmap = require('enmap');
+
+// Create giveaways table
+const giveawayDB = new Enmap({name: "giveaways"})
+
+const { GiveawaysManager } = require('discord-giveaways');
+const GiveawayManagerWithOwnDatabase = class extends GiveawaysManager {
+
+    // This function is called when the manager needs to get all the giveaway stored in the database.
+    async getAllGiveaways(){
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:18]:
```suggestion
    async getAllGiveaways() {
```

File: examples/custom-databases/enmap.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        return giveawayDB.fetchEverything().array();
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:19]:
```suggestion
        // Get all the giveaway in the database
        return giveawayDB.fetchEverything().array();
```

File: examples/custom-databases/enmap.js [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    // This function is called when a giveaway needs to be saved in the database (when a giveaway is created or when a giveaway is edited).
+    async saveGiveaway(messageID, giveawayData){
+        giveawayDB.set(messageID, giveawayData);
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:24]:
```suggestion
        // Add the new one
        giveawayDB.set(messageID, giveawayData);
```

**Review by Nico105 — CHANGES_REQUESTED** [review]
Beautifying 😁 _So that every example is similar and commented_

**Comment by Nico105** [comment]
I would also suggest that you add a link to your example inside of the README.

**Review by Snowflake107 — CHANGES_REQUESTED** [review]
Use **[Enmap#ensure()](https://enmap.evie.dev/api#enmap-ensure-key-defaultvalue-path)**

**Comment by Nico105** [comment]
That's what I also thought in the first place but I mean the `giveawaDB` itself is a "collection" so we do not have to set it to [] right, and so why would we have to ensure then?
Or am I assuming wrong?

**Comment by Snowflake107** [comment]
why dont u make individual document instead of just one single array? You can manage stuff easily (as discord-giveaways is giving u full access to write custom manager)

File: examples/custom-databases/enmap.js [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    settings = {
+        prefix: 'g!',
+        token: 'Your Discord Bot Token'
+    };
+
+// Load Enmap
+const Enmap = require('enmap');
+
+// Create giveaways table
+const giveawayDB = new Enmap({ name: 'giveaways' })
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:12]:
```suggestion
const db = new Enmap({ name: 'giveaways' })
db.ensure('giveaways', [])
```

File: examples/custom-databases/enmap.js [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+const { GiveawaysManager } = require('discord-giveaways');
+const GiveawayManagerWithOwnDatabase = class extends GiveawaysManager {
+    // This function is called when the manager needs to get all the giveaways stored in the database.
+    async getAllGiveaways() {
+        // Get all the giveaway in the database
+        return giveawayDB.fetchEverything().array();
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:19]:
```suggestion
        return db.get('giveaways');
```

File: examples/custom-databases/enmap.js [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    // This function is called when a giveaway needs to be saved in the database (when a giveaway is created or when a giveaway is edited).
+    async saveGiveaway(messageID, giveawayData) {
+        // Add the new one
+        giveawayDB.set(messageID, giveawayData);
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:25]:
```suggestion
        db.push('giveaways', { giveawayData });
```

File: examples/custom-databases/enmap.js [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        // Don't forget to return something!
+        return true;
+    }
+
+    // This function is called when a giveaway needs to be edited in the database.
+    async editGiveaway(messageID, giveawayData) {
+        giveawayDB.set(messageID, giveawayData);
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:32]:
```suggestion
        // Gets all the current giveaways
        const giveaways = db.get('giveaways');
        // Remove the old giveaway from the current giveaways ID
        const newGiveawaysArray = giveaways.filter((giveaway) => giveaway.messageID !== messageID);
        // Push the new giveaway to the array
        newGiveawaysArray.push(giveawayData);
        // Save the updated array
        db.set('giveaways', newGiveawaysArray);
```

File: examples/custom-databases/enmap.js [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+        // Don't forget to return something!
+        return true;
+    }
+
+    // This function is called when a giveaway needs to be deleted from the database.
+    async deleteGiveaway(messageID) {
+        giveawayDB.delete(messageID);
```
Comment (Nico105) [inline examples/custom-databases/enmap.js:39]:
```suggestion
        // Remove the giveaway from the array
        const newGiveawaysArray = db.get('giveaways').filter((giveaway) => giveaway.messageID !== messageID);
        // Save the updated array
        db.set('giveaways', newGiveawaysArray);
        // Don't forget to return something!
        return true;
```

**Review by Nico105 — COMMENTED** [review]
Like that? @Snowflake107

**Comment by GreepTheSheep** [comment]
@Snowflake107 It's possible. But I want to base myself as for MySQL ie to create a new value when there is a new giveaway. Then I put it all together with `.array()`

**Comment by Snowflake107** [comment]
👍

**Review by Androz2091 — APPROVED** [review]
Thank you! 👍

**Pull request merged**